### DISCUSSION TITLE
有什么配置可以让光标快速从括号里移动到右括号后吗？

### DISCUSSION THREAD

**Discussion by lilhammerfun** [neoclide/coc.nvim#4937] [Q&A] [answered]
我的意思是，当我选择了某项选中项后，比如说下图的函数(宏)，我的光标会自动移动到括号里，方便我写入参，但是我想在写完入参后，也很方便地将光标移动到右括号后边，方便我写statement的分号。简而言之，就是实现IDE工具中类似tab键的功能，光标会智能地选择执行缩进还是移动到右括号后。有什么配置可以做到吗？
![image](https://github.com/neoclide/coc.nvim/assets/113404700/bcea66d9-90b0-41bb-bbf7-a9806a0e1f85)

**Comment by fannheyward** [comment, marked as answer]
如果补全项是 snippet，可以用 `<C-j>` (可以通过 `:h coc_snippet_next` 修改快捷键）跳到下一个位置，如果不是 snippet，没办法直接跳出。

**Reply by lilhammerfun** [reply]
好吧，那确实没办法了，这样真的很不方便，需要来回切换模式。

**Comment by lilhammerfun** [comment]
想到了一个可行的方案，我去看了auto pairs的文档，如果我们额外安装一个auto pairs插件，我们就可以保持在insert模式将光标从括号内移到括号外，当然除了括号那些""什么的也行，具体看auto pairs文档吧。

**Comment by lilhammerfun** [comment]
我想到一个绝妙的方案解决这个问题，这个想法基于`coc.txt`源文件中的这一段内容对我的启发：
```vim
Map <tab> for trigger completion, completion confirm, snippet expand and jump 
like VSCode:

  inoremap <silent><expr> <TAB>
    \ coc#pum#visible() ? coc#_select_confirm() :
    \ coc#expandableOrJumpable() ?
    \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
    \ CheckBackspace() ? "\<TAB>" :
    \ coc#refresh()

  function! CheckBackspace() abort
    let col = col('.') - 1
    return !col || getline('.')[col - 1]  =~# '\s'
  endfunction

  let g:coc_snippet_next = '<tab>'
```

于是我写了一个判断下一个字符是否属于成对符号之一的函数：
```vim
function! NextCharIsPair() abort
  let col = col('.') - 1 
  let l:next_char = getline('.')[col]
  return l:next_char =~# ')\|]\|}\|>\|''\|"\|`'
endfunction
```
当符合这种情况时，我们使用Tab键执行一个向右->的操作来跳出成对符号内部：
```vim
inoremap <silent><expr> <Tab>
  \ coc#pum#visible() ? coc#_select_confirm() :
  \ coc#expandableOrJumpable() ?
  \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
  \ NextCharIsPair() ? "\<Right>" :  " 向右跳出成对符号内部
  \ CheckBackspace() ? "\<Tab>" :
  \ coc#refresh()
```

然后我把上面的配置放到了`coc.vim`文件中，使其生效，当然不能忘了安装`coc-snippets`，经过我的测试，所有的这些成对符号我都能在不切换模式的前提下（也即保持插入模式）成功使用Tab键直接跳出，这样确实更像IDE的Tab键功能了。

@fannheyward 这个方案您觉得可行吗？ 如果可以的话，我可以提交一个PR吗？这也是我第一次尝试对开源项目做贡献，如果尽到绵薄之力的话也会感到很开心的！我认为这个方案可以在`coc.txt`中被添加或者在`README.md`中的Example Vim configuration中被提及。

**Reply by fannheyward** [reply]
赞！可以添加到 FAQ 里 https://github.com/neoclide/coc.nvim/wiki/F.A.Q

**Reply by lilhammerfun** [reply]
wiki已修改，您看看修改结果可以吗？ 这个没有审批直接通过了？

另外，我其实还是觉得这个功能虽小，但是对使用vim编码效率的提升是有很大帮助的，仅仅在FAQ里提及，我觉得有些低估了这个功能带来的效益，这一点是我不太理解的地方。
